### PR TITLE
Cleanup DSharpPlus.Rest

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
Splits the `DSharpPlus.Rest` project into multiple files, cleaning up code and adding some documentation.

# Notes
Per usual, entirely untested. Requesting @InFTord's review for incorrect/missing docs or parameters.

### CODE REVIEW

File: DSharpPlus.Rest/DiscordRestClient.Guild.cs [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+    /// <param name="code">The code of the template to delete.</param>
+    /// <returns>The deleted template.</returns>
+    public Task<DiscordGuildTemplate> DeleteGuildTemplateAsync(ulong guildId, string code)
+        => ApiClient.DeleteGuildTemplateAsync(guildId, code);
+
+    /// <summary>
+    /// Gets a guild's welcome screen.
+    /// </summary>
+    /// <returns>The guild's welcome screen object.</returns>
+    public Task<DiscordGuildWelcomeScreen> GetGuildWelcomeScreenAsync(ulong guildId)
```
Comment (InFTord) [inline DSharpPlus.Rest/DiscordRestClient.Guild.cs:395]: param isn't documented

**Review by InFTord — COMMENTED** [review]
one small nitpick, but overall it's good

File: DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    public Task<ThreadQueryResult> ListActiveThreadAsync(ulong guildId)
+        => ApiClient.ListActiveThreadsAsync(guildId);
+
+    /// <summary>
+    /// Gets the threads that are public and archived for a channel.
+    /// </summary>
+    /// <param name="guildId">The ID of the guild.</param>
+    /// <param name="channelId">The ID of the channel.</param>
+    /// <param name="before">Date to filter by.</param>
+    /// <param name="limit">Limit.</param>
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs:106]: maybe elaborate a little more

File: DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    public Task<ThreadQueryResult> ListPublicArchivedThreadsAsync(ulong guildId, ulong channelId, DateTimeOffset? before = null, int limit = 0)
+       => ApiClient.ListPublicArchivedThreadsAsync(guildId, channelId, before?.ToString("o"), limit);
+
+    /// <summary>
+    /// Gets the threads that are public and archived for a channel.
+    /// </summary>
+    /// <param name="guildId">The ID of the guild.</param>
+    /// <param name="channelId">The ID of the channel.</param>
+    /// <param name="before">Date to filter by.</param>
+    /// <param name="limit">Limit.</param>
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs:116]: as above

File: DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    public Task<ThreadQueryResult> ListPrivateArchivedThreadAsync(ulong guildId, ulong channelId, DateTimeOffset? before = null, int limit = 0)
+       => ApiClient.ListPrivateArchivedThreadsAsync(guildId, channelId, before?.ToString("o"), limit);
+
+    /// <summary>
+    /// Gets the private archived threads the user has joined for a channel.
+    /// </summary>
+    /// <param name="guildId">The ID of the guild.</param>
+    /// <param name="channelId">The ID of the channel.</param>
+    /// <param name="before">Date to filter by.</param>
+    /// <param name="limit">Limit.</param>
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.Threads.cs:126]: as above

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+        bool? nsfw,
+        Optional<int?> perUserRateLimit,
+        VideoQualityMode? qualityMode,
+        int? position,
+        string reason,
+        AutoArchiveDuration? defaultAutoArchiveDuration = null,
+        DefaultReaction? defaultReactionEmoji = null,
+        IEnumerable<DiscordForumTagBuilder>? availableTags = null,
+        DefaultSortOrder? defaultSortOrder = null
+    ) => type switch
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.cs:99]: please put this in its own block, not an expression body

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+        IEnumerable<DiscordOverwriteBuilder> permissionOverwrites,
+        string reason,
+        Optional<ChannelFlags> flags,
+        IEnumerable<DiscordForumTagBuilder>? availableTags,
+        Optional<AutoArchiveDuration?> defaultAutoArchiveDuration,
+        Optional<DefaultReaction?> defaultReactionEmoji,
+        Optional<int> defaultPerUserRatelimit,
+        Optional<DefaultSortOrder?> defaultSortOrder,
+        Optional<DefaultForumLayout> defaultForumLayout
+    ) => ApiClient.ModifyChannelAsync
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.cs:175]: same here

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+        ulong channelId,
+        int maxAge,
+        int maxUses,
+        bool temporary,
+        bool unique,
+        string reason,
+        InviteTargetType? targetType = null,
+        ulong? targetUserId = null,
+        ulong? targetApplicationId = null
+    ) => ApiClient.CreateChannelInviteAsync
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Channels.cs:289]: same here

File: DSharpPlus.Rest/DiscordRestClient.Guild.cs [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+    /// <param name="defaultMessageNotifications">New guild's default message notification level</param>
+    /// <param name="systemChannelFlags">New guild's system channel flags</param>
+    public Task<DiscordGuild> CreateGuildAsync(
+        string name,
+        string regionId,
+        string iconBase64,
+        VerificationLevel? verificationLevel,
+        DefaultMessageNotifications? defaultMessageNotifications,
+        SystemChannelFlags? systemChannelFlags
+    ) => ApiClient.CreateGuildAsync(name, regionId, iconBase64, verificationLevel, defaultMessageNotifications, systemChannelFlags);
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Guild.cs:59]: formatting

File: DSharpPlus.Rest/DiscordRestClient.Guild.cs [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+        Optional<string> banner,
+        Optional<string> description,
+        Optional<string> discoverySplash,
+        Optional<IEnumerable<string>> features,
+        Optional<string> preferredLocale,
+        Optional<ulong?> publicUpdatesChannelId,
+        Optional<ulong?> rulesChannelId,
+        Optional<SystemChannelFlags> systemChannelFlags,
+        string reason
+    ) => ApiClient.ModifyGuildAsync(
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Guild.cs:134]: as above

File: DSharpPlus.Rest/DiscordRestClient.Members.cs [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    /// <param name="deafened">Whether this user should be deafened on join</param>
+    public Task<DiscordMember> AddGuildMemberAsync(
+        ulong guildId,
+        ulong userId,
+        string accessToken,
+        string nick,
+        IEnumerable<DiscordRole> roles,
+        bool muted,
+        bool deafened
+    ) => ApiClient.AddGuildMemberAsync(guildId, userId, accessToken, nick, roles, muted, deafened);
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Members.cs:55]: formatting

File: DSharpPlus.Rest/DiscordRestClient.Members.cs [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        ulong guildId,
+        ulong userId,
+        Optional<string> nick,
+        Optional<IEnumerable<ulong>> roleIds,
+        Optional<bool> mute,
+        Optional<bool> deaf,
+        Optional<ulong?> voiceChannelId,
+        Optional<DateTimeOffset?> communicationDisabledUntil,
+        string reason
+    ) => ApiClient.ModifyGuildMemberAsync(guildId, userId, nick, roleIds, mute, deaf, voiceChannelId, communicationDisabledUntil, reason);
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Members.cs:150]: formatting

File: DSharpPlus.Rest/DiscordRestClient.Roles.cs [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+        ulong roleId,
+        string name,
+        Permissions? permissions,
+        DiscordColor? color,
+        bool? hoist,
+        bool? mentionable,
+        string reason,
+        Stream icon,
+        DiscordEmoji emoji
+    ) => ApiClient.ModifyGuildRoleAsync
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.Roles.cs:82]: formatting

**Review by akiraveliara — CHANGES_REQUESTED** [review]
the formatting is very inconsistent, both with itself and with allman style. i've left comments on the most egregious violations i've caught, but i implore you to take a look over all of it - there's a fair few more.
i've also left a nitpick or two but those aren't super important

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 13]
```diff
@@ -21,2279 +21,51 @@
...
+    public override IReadOnlyDictionary<ulong, DiscordGuild> Guilds => _guilds;
+    internal Dictionary<ulong, DiscordGuild> _guilds = new();
+    private bool _disposed;
+
+    public DiscordRestClient(DiscordConfiguration config) : base(config) => _disposed = false;
+
+    /// <summary>
+    /// Initializes cache
+    /// </summary>
+    /// <returns></returns>
```
Comment (InFTord) [inline DSharpPlus.Rest/DiscordRestClient.cs:44]: returns what?

File: DSharpPlus.Rest/DiscordRestClient.ApplicationCommands.cs [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+    /// <param name="type">The type of the response.</param>
+    /// <param name="builder">The data, if any, to send.</param>
+    public Task CreateInteractionResponseAsync(ulong interactionId, string interactionToken, InteractionResponseType type, DiscordInteractionResponseBuilder? builder = null)
+        => ApiClient.CreateInteractionResponseAsync(interactionId, interactionToken, type, builder);
+
+    /// <summary>
+    /// Gets the original interaction response.
+    /// </summary>
+    /// <returns>The original message that was sent. This <b>does not work on ephemeral messages.</b></returns>
+    public Task<DiscordMessage> GetOriginalInteractionResponseAsync(string interactionToken)
```
Comment (InFTord) [inline DSharpPlus.Rest/DiscordRestClient.ApplicationCommands.cs:179]: param not documented

File: DSharpPlus.Rest/DiscordRestClient.ApplicationCommands.cs [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    /// <summary>
+    /// Batch edits permissions for a application command in a guild.
+    /// </summary>
+    /// <param name="guildId">The guild ID.</param>
+    /// <param name="permissions">The list of permissions to use.</param>
+    /// <returns>A list of edited permissions.</returns>
+    public Task<IReadOnlyList<DiscordGuildApplicationCommandPermissions>> BatchEditApplicationCommandPermissionsAsync(ulong guildId, IEnumerable<DiscordGuildApplicationCommandPermissions> permissions)
+        => ApiClient.BatchEditApplicationCommandPermissionsAsync(CurrentApplication.Id, guildId, permissions);
+
+    public Task<DiscordMessage> GetFollowupMessageAsync(string interactionToken, ulong messageId)
```
Comment (InFTord) [inline DSharpPlus.Rest/DiscordRestClient.ApplicationCommands.cs:272]: not documented at all

File: DSharpPlus.Rest/DiscordRestClient.Users.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    /// </summary>
+    /// <param name="username">New username</param>
+    /// <param name="avatarBase64">New avatar (base64)</param>
+    public async Task<DiscordUser> ModifyCurrentUserAsync(string? username, string? avatarBase64)
+        => new DiscordUser(await ApiClient.ModifyCurrentUserAsync(username, avatarBase64)) { Discord = this };
+
+    /// <summary>
+    /// Modifies current user
+    /// </summary>
+    /// <param name="username">username</param>
```
Comment (InFTord) [inline DSharpPlus.Rest/DiscordRestClient.Users.cs:58]: add more info about what are those params

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs (hunk 5 above)
Comment (OoLunar) [reply DSharpPlus.Rest/DiscordRestClient.Channels.cs:99]: The `.editorconfig` will cause a compiler warning, informing us to make it an expression. What is your proposed solution?

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs (hunk 6 above)
Comment (OoLunar) [reply DSharpPlus.Rest/DiscordRestClient.Channels.cs:175]: x2

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs (hunk 7 above)
Comment (OoLunar) [reply DSharpPlus.Rest/DiscordRestClient.Channels.cs:289]: x2

File: DSharpPlus.Rest/DiscordRestClient.Guild.cs (hunk 9 above)
Comment (OoLunar) [reply DSharpPlus.Rest/DiscordRestClient.Guild.cs:141]: x2

File: DSharpPlus.Rest/DiscordRestClient.Roles.cs (hunk 12 above)
Comment (OoLunar) [reply DSharpPlus.Rest/DiscordRestClient.Roles.cs]: x2

File: DSharpPlus.Rest/DiscordRestClient.Channels.cs (hunk 5 above)
Comment (akiraveliara) [reply DSharpPlus.Rest/DiscordRestClient.Channels.cs:99]: change it to only prefer expressions on single lines?

**Comment by Plerx2493** [comment]
I think this is superseded by #1624 ?

**Comment by akiraveliara** [comment]
no, but it would at least need to be modified for #1624

**Comment by OoLunar** [comment]
Closing as it's blocked by #1647. When that PR merges, I will attempt this PR once more.

**Pull request closed without merging**